feat: Allow usage of `print()` within `safe_exec()`

akhilnarang opened the pull request:
This resolves #22582

`print()` can now be used within `safe_exec()` calls - server scripts, system console, etc.

For system console, it shows normally in output
![image](https://github.com/frappe/frappe/assets/10119037/35a38cce-3fb6-4471-a39f-a4ecedd1e7b0)

For server scripts, a new field has been added `Debug Log`

![image](https://github.com/frappe/frappe/assets/10119037/44d8253f-173c-4b1a-b8bb-37a452285659)

For doc events and API, the printed values are part of debug log.

![image](https://github.com/frappe/frappe/assets/9079960/e5418615-ae26-4eef-84b5-811a5f2448b0)



[no-docs]

akhilnarang committed:
feat(safe_exec): allow usage of `print()`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(system_console): update description to mention `print()` instead of `log()`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/utils/safe_exec.py:
```diff
@@ -60,6 +61,16 @@ def check_name(self, node, name, *args, **kwargs):
+	def _call_print(self, *objects, **kwargs):
+		output = io.StringIO()
+		print(*objects, file=output, **kwargs)
+		frappe.log(output.getvalue().strip())
```
This doesn't work without enabling debug logging using `logging` key in site config. 

Can we ignore it and do it always in case of safe exec prints? :thinking:

akhilnarang committed:
feat: unconditionally add debug logs to response if present

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(safe_exec): add in a test for running `print()` within safe_exec

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/__init__.py:
```diff
@@ -155,6 +155,8 @@ def set_user_lang(user: str, user_language: str | None = None) -> None:
 debug_log = local("debug_log")
 message_log = local("message_log")
 
+force_debug_logs = local("force_debug_logs")
```
`frappe.log(..., force=True)` will be better?

This way we can ignore logging param in site config?

akhilnarang commented:
Current state works fine, just has this warning

```
frappe-bench/env/lib/python3.11/site-packages/RestrictedPython/compile.py:204: SyntaxWarning: Line None: Prints, but never reads 'printed' variable.
warnings.warn(
```

Comes from https://github.com/zopefoundation/RestrictedPython/blob/master/src/RestrictedPython/transformer.py#L478-L509
https://github.com/zopefoundation/RestrictedPython/blob/master/src/RestrictedPython/transformer.py#L630

akhilnarang committed:
fix(safe_exec): ignore warning

RestrictedPython warns us if we call `print()` don't use their `printed` variable

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: store debug logs from scheduled jobs

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush committed:
Merge branch 'develop' into restrictedpython-print

ankush committed:
fix: avoid ignoring warnings, disabled in prod anyway

ankush committed:
chore: remove unnecessary logging

This can be moved to level 2 when required
